### ISSUE TITLE
New Containers aren't healthy after SpiceDBCluster spec.config changes. Deployments blocked. `service unhealthy (responded with "NOT_SERVING")`

### ISSUE THREAD

**Issue by jwmay2012** [authzed/spicedb-operator#174]
If I make any change to a `SpiceDBCluster`, the new containers will not become healthy, and not take over the old deployment.
Whether this deployment requires a new ReplicaSet or not. 

`kubectl -n spicedb get spicedbcluster/spicedb`
```
NAME                                 AGE     CHANNEL   DESIRED   CURRENT   WARNINGS   MIGRATING   UPDATING   INVALID   PAUSED
spicedbcluster.authzed.com/spicedb   5d22h                       v1.18.0                          True
```

`kubectl -n spicedb describe spicedbcluster/spicedb`
```
...
Status:
  Conditions:
    Last Transition Time:  2023-03-27T19:35:54Z
    Message:               Waiting for deployment to be available: 1/2 available, 1/2 ready, 2/2 updated, 8/8 generation.
    Reason:                WaitingForDeploymentAvailability
    Status:                True
    Type:                  RollingDeployment
...
```

`kubectl get po -n spicedb`
```
NAME                                READY   STATUS             RESTARTS         AGE
spicedb-operator-7968686549-8d4nn   1/1     Running            0                3d
spicedb-spicedb-5984bd67c-r4vct     0/1     CrashLoopBackOff   6 (2m13s ago)    14m
spicedb-spicedb-5984bd67c-x8c2s     1/1     Running            69 (2d19h ago)   3d
```

`kubectl -n spicedb describe pod/spicedb-spicedb-5984bd67c-r4vct`

```
Events:
  Type     Reason     Age   From               Message
  ----     ------     ----  ----               -------
  Normal   Scheduled  84s   default-scheduler  Successfully assigned spicedb/spicedb-spicedb-5984bd67c-r4vct to us1c-b-gce-gp-n2c2m8-a-b-9mmxp
  Normal   Pulling    84s   kubelet            Pulling image "ghcr.io/authzed/spicedb:v1.18.0"
  Normal   Pulled     80s   kubelet            Successfully pulled image "ghcr.io/authzed/spicedb:v1.18.0" in 3.660331808s (3.660354092s including waiting)
  Normal   Created    80s   kubelet            Created container spicedb-spicedb
  Normal   Started    80s   kubelet            Started container spicedb-spicedb
  Warning  Unhealthy  78s   kubelet            Readiness probe failed: parsed options:
> addr=localhost:50051 conn_timeout=1s rpc_timeout=1s
> tls=true
  > no-verify=true
  > ca-cert=
  > client-cert=
  > client-key=
  > server-name=
> alts=false
> spiffe=false
establishing connection
timeout: failed to connect service "localhost:50051" within 1s
  Warning  Unhealthy  78s  kubelet  Readiness probe failed: parsed options:
> addr=localhost:50051 conn_timeout=1s rpc_timeout=1s
> tls=true
  > no-verify=true
  > ca-cert=
  > client-cert=
  > client-key=
  > server-name=
> alts=false
> spiffe=false
establishing connection
connection established (took 8.625008ms)
service unhealthy (responded with "NOT_SERVING")
  Warning  Unhealthy  77s  kubelet  Readiness probe failed: parsed options:
> addr=localhost:50051 conn_timeout=1s rpc_timeout=1s
> tls=true
  > no-verify=true
  > ca-cert=
  > client-cert=
  > client-key=
  > server-name=
> alts=false
> spiffe=false
establishing connection
connection established (took 7.656165ms)
service unhealthy (responded with "NOT_SERVING")
...repeats
```

`kubectl -n spicedb logs pod/spicedb-spicedb-5984bd67c-r4vct`
```
{"level":"info","format":"auto","log_level":"info","provider":"zerolog","async":false,"time":"2023-03-27T19:41:06Z","message":"configured logging"}
{"level":"info","v":0,"provider":"none","endpoint":"","service":"spicedb","insecure":false,"sampleRatio":0.2,"time":"2023-03-27T19:41:06Z","message":"configured opentelemetry tracing"}
{"level":"info","latest-released-version":"v1.18.0","time":"2023-03-27T19:41:07Z","message":"this is the latest released version of SpiceDB"}
{"level":"info","time":"2023-03-27T19:41:07Z","message":"using postgres datastore engine"}
{"level":"warn","time":"2023-03-27T19:41:07Z","message":"watch API disabled, postgres must be run with track_commit_timestamp=on"}
{"level":"info","initialSlowRequest":"10ms","maxRequests":1000000,"hedgingQuantile":0.95,"time":"2023-03-27T19:41:07Z","message":"request hedging enabled"}
{"level":"info","interval":180000,"time":"2023-03-27T19:41:07Z","message":"datastore garbage collection worker started"}
{"level":"info","maxCost":"16 MiB","numCounters":1000,"defaultTTL":0,"time":"2023-03-27T19:41:07Z","message":"configured namespace cache"}
{"level":"info","maxCost":"311 MiB","numCounters":10000,"defaultTTL":10000,"time":"2023-03-27T19:41:07Z","message":"configured dispatch cache"}
{"level":"info","check-permission":50,"lookup-resources":50,"lookup-subjects":50,"reachable-resources":50,"time":"2023-03-27T19:41:07Z","message":"configured dispatch concurrency limits"}
{"level":"info","maxCost":"726 MiB","numCounters":100000,"defaultTTL":10000,"time":"2023-03-27T19:41:07Z","message":"configured cluster dispatch cache"}
{"level":"info","addr":":50053","network":"tcp","service":"dispatch-cluster","workers":0,"insecure":false,"time":"2023-03-27T19:41:07Z","message":"grpc server started serving"}
{"level":"warn","reason":"","time":"2023-03-27T19:41:07Z","message":"watch api disabled; underlying datastore does not support it"}
{"level":"info","addr":":50051","network":"tcp","service":"grpc","workers":0,"insecure":false,"time":"2023-03-27T19:41:07Z","message":"grpc server started serving"}
{"level":"info","addr":":9090","service":"metrics","insecure":true,"time":"2023-03-27T19:41:07Z","message":"http server started serving"}
{"level":"info","addr":":8080","prefix":"dashboard","insecure":false,"time":"2023-03-27T19:41:07Z","message":"http server started serving"}
{"level":"info","time":"2023-03-27T19:41:07Z","message":"telemetry disabled"}
{"level":"info","grpc.component":"server","grpc.method":"Check","grpc.method_type":"unary","grpc.service":"grpc.health.v1.Health","peer.address":"[::1]:34376","protocol":"grpc","requestID":"032b0eb4fbf6705d6e4b61fd06f976ae","grpc.request.deadline":"2023-03-27T19:41:08Z","grpc.start_time":"2023-03-27T19:41:07Z","grpc.code":"OK","grpc.time_ms":"0.116","time":"2023-03-27T19:41:07Z","message":"finished call"}
{"level":"info","grpc.component":"server","grpc.method":"Check","grpc.method_type":"unary","grpc.service":"grpc.health.v1.Health","peer.address":"[::1]:34382","protocol":"grpc","requestID":"2e8efe06bd28793a95262ff099d43126","grpc.request.deadline":"2023-03-27T19:41:09Z","grpc.start_time":"2023-03-27T19:41:08Z","grpc.code":"OK","grpc.time_ms":"0.103","time":"2023-03-27T19:41:08Z","message":"finished call"}
{"level":"info","grpc.component":"server","grpc.method":"Check","grpc.method_type":"unary","grpc.service":"grpc.health.v1.Health","peer.address":"[::1]:51758","protocol":"grpc","requestID":"2aad4920db16e32bbd49ac687419460a","grpc.request.deadline":"2023-03-27T19:41:17Z","grpc.start_time":"2023-03-27T19:41:16Z","grpc.code":"OK","grpc.time_ms":"0.077","time":"2023-03-27T19:41:16Z","message":"finished call"}
...repeats
```

Seems like changes can be anything. Toggling `telemetry` on and off. Increasing `replicas` by 1. 
All results in the new containers failing the Readiness probe  `service unhealthy (responded with "NOT_SERVING")`.

Any insight as to what's going on here?

**Comment by jwmay2012** [comment]
If I delete everything, spicedb and the operator, and deploy again with any of the changes, it starts up fine. 
So it seems to work fine starting up by itself. 
Even if I start it with `replicas: 2`, they both startup clean.

BUT, if I `delete` one of the replicas, the one that gets started to replace it hits the same `service unhealthy (responded with "NOT_SERVING")` issue.
So if both pods are started up simultaneously, it's fine. But if it's one, then another, there's an issue?

```
apiVersion: authzed.com/v1alpha1
kind: SpiceDBCluster
metadata:
  name: spicedb
  namespace: spicedb
  annotations:
    kapp.k14s.io/change-group: spicedb.acreops.com/cluster
spec:
  config:
    datastoreEngine: postgres
    httpEnabled: true
    replicas: 2
    tlsSecretName: spicedb-grpc-tls
    datastoreTLSSecretName: spicedb-grpc-tls
    telemetryEndpoint: ""
  secretName: spicedb-config
  ```

**Comment by ecordell** [comment]
@jwmay2012 I suspect you're hitting this issue: https://github.com/authzed/spicedb/issues/814

What's happening is that on initial startup, SpiceDB is reporting all nodes healthy because they don't see each other for dispatch (spicedb-to-spicedb traffic) yet. Once you make a change and the pods roll, the new pod _does_ see the other one for dispatch, and notices that there's a connection error. The check only happens on startup.

If you set:
```yaml
spec:
  config:
    logLevel: debug
```
I suspect you'll see a log message in the failing pod saying dispatch is unavailable.

Is there anything unusual about the networking configuration in your cluster? Rules that might prevent the pods from talking to each other?

**Comment by jwmay2012** [comment]
```
{"level":"debug","time":"2023-03-27T20:21:40Z","message":"checking if datastore and dispatcher are ready"}
{"level":"debug","datastoreReady":true,"dispatchReady":false,"time":"2023-03-27T20:21:40Z","message":"completed dispatcher and datastore readiness checks"}
```
That's it! There's dispatcher issues. I'll look into this and report back. Thanks!

**Comment by jwmay2012** [comment]
Its probably due to bad or missing TLS certs on dispatcher communication. I've got certs setup for grpc which apparently sets them for the dispatcher too. 
I've fixed this by disabling the dispatcher for now until I can generate proper certs and set `dispatchUpstreamCASecretName`.

But the linked issue with the race condition is certainly why a quick startup with TLS still works. 

Closing. Thanks for your help!
